engine/src/caching_client.rs

```diff
@@ -0,0 +1,224 @@
+	time::timeout,
+};
+
+const MAX_DURATION: Duration = Duration::new(30, 0);
+
+type PinBoxedFuture<ResultType> =
+	Pin<Box<dyn Future<Output = Result<ResultType, anyhow::Error>> + Send>>;
+#[derive(Clone)]
+pub struct CachingRequest<Key, ResultType, Client> {
+	request_sender: mpsc::UnboundedSender<(
+		PinBoxedFuture<ResultType>,
+		Key,
+		oneshot::Sender<Result<ResultType, anyhow::Error>>,
+	)>,
+	client: Client,
+}
+
+impl<
+		Key: Eq + Hash + Clone + Send + 'static,
+		ResultType: Clone + Send + 'static,
+		Client: Clone + Send + Sync + 'static,
+	> CachingRequest<Key, ResultType, Client>
+{
+	pub fn new(scope: &Scope<'_, anyhow::Error>, client: Client) -> (Self, mpsc::Sender<()>) {
+		let (request_sender, mut request_receiver) = mpsc::unbounded_channel::<(
+			PinBoxedFuture<ResultType>,
+			Key,
+			oneshot::Sender<Result<ResultType, anyhow::Error>>,
+		)>();
+		let (cache_invalidation_sender, mut cache_invalidation_receiver) = mpsc::channel::<()>(1);
+		scope.spawn_weak({
+			task_scope::task_scope(|scope| {
+				async move {
+					let mut cache: HashMap<Key, ResultType> = HashMap::new();
+					let mut in_flight: HashMap<
+						Key,
+						Vec<oneshot::Sender<Result<ResultType, anyhow::Error>>>,
+					> = HashMap::default();
+					let (cache_sender, mut cache_receiver) =
+						mpsc::unbounded_channel::<(Key, Result<ResultType, anyhow::Error>)>();
+					loop {
+						tokio::select! {
+							biased;
+							Some(()) = cache_invalidation_receiver.recv() => {
+								// We only clear the cache, we don't want to cancel in_flight requests, this way we are sure that every request will eventually complete
+								// regardless of the time it takes to complete
+								cache.clear();
+							},
+							Some((key, value)) = cache_receiver.recv() => {
+								if let Ok(ref value) = value {
+									cache.insert(key.clone(), value.clone());
+								}
+								if let Some(senders) = in_flight.remove(&key) {
+									for sender in senders {
+										let to_send = match &value {
+											Ok(val) => Ok(val.clone()),
+											Err(e) => Err(anyhow::anyhow!(e.to_string())),
+										};
+										let _ = sender.send(to_send);
+									}
+								}
+							},
+							Some((future, request_key, sender)) = request_receiver.recv() => {
+								if let Some(value) = cache.get(&request_key)  {
+										let _ = sender.send(Ok(value.clone()));
+								} else if let Some(result_senders) = in_flight.get_mut(&request_key) {
+									result_senders.push(sender);
+								} else {
+									let sender_result = cache_sender.clone();
+									in_flight.insert(request_key.clone(), vec![sender]);
+									scope.spawn(async move {
+										let result = future.await;
+										let _ = sender_result.send((request_key, result));
+										Ok(())
+									})
+								}
+
+							},
+						}
+					}
+				}
+				.boxed()
+			})
+		});
+
+		(Self { request_sender, client }, cache_invalidation_sender)
+	}
+
+	pub(crate) async fn get(
```

Comment:
nit: we could probably use a slightly more descriptive name like `get_or_fetch` to imply there are two modes that could be happening when this function is called